### PR TITLE
Add sbi wrapper class

### PR DESCRIPTION
**Pull request by antelk**
Attempt to resolve the issue #44.

The class Inferencer is fully functional and can be tested via `hh_sbi.py` file in examples.
It supports both multi-round inference and multi-input/output traces.

Even though the class is ready to use, the feature metric is not yet fully supported.
This actual feature metric should be supported by resolving the issue #48.

All additional changes are more or less stylistic improvements.
`sbi` is added in the requirements file and in the `setup.py` as the optional dependency.

### CODE REVIEW

File: brian2modelfitting/inferencer.py [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+            # extract the data and make adjustments for `sbi`
+            theta, simulator = self.generate_training_data(n_samples, proposal)
+            theta = torch.tensor(theta, dtype=torch.float32)
+            obs = simulator.statemonitor.recorded_variables
+            x = []
+            for ov in self.output_var:
+                x_val = obs[ov].get_value_with_unit()
+                x_dim = get_dimensions(obs[ov])
+                features = self.extract_features(x_val, x_dim)
+                x.append(features)
```
Comment (mstimberg) [inline brian2modelfitting/inferencer.py:478]: See above comment about units, e.g. here you could simply use `get_value()`.

File: brian2modelfitting/inferencer.py [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+            de = inference.append_simulations(theta, x, proposal).train()
+
+            # use the density estimator to build the posterior
+            posterior = inference.build_posterior(de)
+
+            # append the current posterior to the list of posteriors
+            posteriors.append(posterior)
+
+            # update the proposal given the observation
+            proposal = posterior.set_default_x(x_o)
```
Comment (mstimberg) [inline brian2modelfitting/inferencer.py:492]: I think it would be good to divide this up a bit further, e.g. have one method to simulate from the priors (or from posteriors from a previous run in the multi-round approach), one method to train the network, and one method to build the posterior. There are several reasons for this: 1) it makes it possible to train a network and then use the trained network on different experimental data (e.g. different cells) 2) it can be useful for debugging/sanity checks (e.g. if half of the simulations from the prior distribution give completely senseless results, it probably makes sense to adapt the prior before starting to train a network and 3) sbi's `train()` and `build_posterior()` methods both take several arguments. For advanced users, it would be good if they could provide them (which would be a mess with the current `Inferencer.train()` method. Having said all that, I think your `Inferencer.train()` method is a good convenience method for the most common use case, so I'd keep it and just have it call out to the more fine-grained methods .

File: examples/hh_sbi.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                 gl=[1e-09 * siemens, 1e-07 * siemens],
+                 g_na=[2e-06 * siemens, 2e-04 * siemens],
+                 g_kd=[6e-07 * siemens, 6e-05 * siemens],
+                 Cm=[0.1 * uF * cm ** -2 * area, 2 * uF * cm ** -2 * area])
+
+# Draw samples from posterior and visualize the results
+labels_params = [r'$\overline{g}_{l}$', r'$\overline{g}_{Na}$',
+                 r'$\overline{g}_{K}$', r'$\overline{C}_{m}$']
+samples = inferencer.sample(1000, viz=True,
+                            labels=labels_params, figsize=(10, 10))
```
Comment (mstimberg) [inline examples/hh_sbi.py:57]: Yet another minor thing: for a quick look how the fitting went, it would be good to introduce a `Inferencer.generate` method (like `Fitter.generate`) that can then be used to visualize the fitted traces in comparison to the experimental ones (like in the other examples).

**Review by mstimberg — COMMENTED** [review]
Everything looking :+1: in general. I left a few specific comments, but nothing major. Regarding the features, I'd think I'd do a short-term solution before dealing with #48 (there are some subtleties we have to take into account, e.g. eFEL often needs stimulus times as an additional argument). This solution would be that the train function (and/or `Inferencer.__init__`) takes a list of functions which each takes as input a 2d-array of traces (samples × inputs) and should give as an output either a 2d array (features × inputs), or a 1d array of shape `(inputs, )` (as convenience for scalar features). These will then be flattened and concatenated before handing over to `sbi`. Note that I did not restrict it to scalar features, I think in some cases it is simpler to write a single function that returns several features instead of asking the user to split them up into multiple functions. The only requirement is that the number of features returned by a function never changes (so something like "spike times" would not be a valid feature, but "min and max membrane potential" would be). In the long run we should certainly check this for a nice error message, but in the short run it is ok if it just fails somewhere in the pipeline with a shape mismatch.

File: brian2modelfitting/utils.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 from brian2 import have_same_dimensions
 from brian2.units.fundamentalunits import Quantity
-from tqdm.autonotebook import tqdm
+from tqdm import tqdm
```
[the remark this answers is not in the document]
Comment (antelk) [reply brian2modelfitting/utils.py:5]: Yes, `autonotebook` looks better but also throws the warning and sometimes fail due to IProgress not found.
Of course, I will change this import as it has been done so far and use `tqdm` from `autonotebook` module.

File: examples/hh_sbi.py (hunk 3 above)
Comment (antelk) [reply examples/hh_sbi.py:57]: Sure thing :)

**Comment by antelk** [comment]
Marcel, thanks for great comments!
We can discuss the issue #48 in greater details in near future.
For now, I'll try to create this list of functions for feature extraction possible.
I guess it would be okay if some of the functions from that list use `eFEL` if the start and end stimulation times are known a priori?

**Comment by mstimberg** [comment]
> For now, I'll try to create this list of functions for feature extraction possible.
> I guess it would be okay if some of the functions from that list use `eFEL` if the start and end stimulation times are known a priori?

So just to avoid misunderstanding, I meant that the inferencer gets something like `features=[fun1, fun2, fun3]` as an input, with `fun1`, etc. being functions mapping traces → features, all independent of `eFEL`. So from the `Inferencer` side we would not offer any functions, all functions would be provided by the user. Concretely, the `hh_sbi.py` example would define a function similar to what is current `Inferencer.extract_features` (which would be removed from `Inferencer`) and pass it as an argument for `features`. In general, the provided functions could certainly be a wrapper around `eFEL`, but the `Inferencer` code should (at this stage) do not need to know that `eFEL` exists, that there is such a thing as a stimulation time, etc.

**Comment by antelk** [comment]
> In general, the provided functions could certainly be a wrapper around `eFEL`, but the `Inferencer` code should (at this stage) do not need to know that `eFEL` exists, that there is such a thing as a stimulation time, etc.

That's exactly what I was thinking. Anyway, thank you for the clarification!

**Comment by antelk** [comment]
Last few commits deal with some minor docstrings issues.
Also `train` method is renamed to `infere`. This was done to be able to refactor the initial train method to 3 additional methods. The reason behind this is stated here:
> I think it would be good to divide this up a bit further, e.g. have one method to simulate from the priors (or from posteriors from a previous run in the multi-round approach), one method to train the network, and one method to build the posterior. There are several reasons for this: 1) it makes it possible to train a network and then use the trained network on different experimental data (e.g. different cells) 2) it can be useful for debugging/sanity checks (e.g. if half of the simulations from the prior distribution give completely senseless results, it probably makes sense to adapt the prior before starting to train a network and 3) sbi's `train()` and `build_posterior()` methods both take several arguments. For advanced users, it would be good if they could provide them (which would be a mess with the current `Inferencer.train()` method. Having said all that, I think your `Inferencer.train()` method is a good convenience method for the most common use case, so I'd keep it and just have it call out to the more fine-grained methods .
This also solves #50.

Advanced users can use each of the methods `init_inference`, `train` and `build_posterior` separately to gain complete control over the inference. Utility method 'infere' achieves similar results but does not offer such a powerful control over `sbi`.

Also, functionality for generating fitted traces by using the inferred parameters is added.

The example is updated according to these changes.

**Review by mstimberg — APPROVED** [review]
Ok, all looks good to me for a merge – I made some minor changes along the way, but nothing really important. We'll need to get into tests and documentation at some point of course, but for now let's continue a bit until the API is settled.

**Comment by antelk** [comment]
Great, thanks!
